# Python Variable Review

In this notebook, some of the code cells are blank and are for you to enter code. Others already have code in them. Make sure you run these as you go, even if it doesn't explicitly say to do so.

## Everything is an object

### Objects in Python

In Python, everything is an object. An object is a single unique unit of information that occupies a chunk of computer memory (the amount of memory depends on the object). 

Objects are generally created by assigning them a name. Like this:

In [3]:
anObject = 42

After you run this cell (run it now!), there is an object in memory that has the name `anOject` assigned to it and the object's value 42.

Objects are like fancy museum display cases. Think about the display case of, for example, the U.S. Declaration of Independence. It has the area beneath clear glass that displays the document itself, but there are also all kinds of fancy bits to it, like lights and sensors and stuff.

Our object's value, 42, is the thing (the noun) that is "on display", which we can see by typing a name currently assigned to it. Type `anObject` in the cell below and run the cell.

In [5]:
anObject

42

The object comes with other built-in nouns and even some verbs too. One of the most fundamental *properties* (nouns) of an object is its *type*. We can get the the type with the `type()` function, which takes one argument, the name of the object. Use the `type()` to get the type of our object in the cell below.

In [6]:
type(anObject)

int

The verbs are accessed by using a period (`.`) after the objects current name and then typing the verb. One of the verbs is getting the bit length – how many "bits" (a contraction of BInary digITS coined by John Tukey) it takes to represent the value in binary notation. Lets, see, for 42... 2, 4, 8, 16, 32, 64, and done! So it should take 6 bits to represent 42 in binary. Let's see using the *method* (verb) `bit_length()`:

In [7]:
anObject.bit_length()

6

Were we right? Try some other numbers!

### Objects and Names

Objects and their names are not the same! Notice that above we said "... *a* name currently assigned..." and "... the object's current name..." A single object can have more than one name.

In [8]:
notAnotherObject = 42

Running that cell does *not* create a second object containing 42. Rather, the "42" object that already exists in memory is simply assigned a second name! 

Another *property* (noun) that all objects have is a unique identification number. The ID number is independent of any names that may be assigned to the object! 

Python has an `id()` function that we can use this to see what names are referring to what objects. Like `type()`, it takes one argument, a name of an object. Try it to get the id number of `anObject`.

In [9]:
id(anObject)

140590022008400

Now use it get the id number of `notAnotherObject`.

In [10]:
id(notAnotherObject)

140590022008400

See? The two names refer to the same object, which saves on memory!

If we do something like this:

In [11]:
anObject = 11

Now an new object storing the value 11 *has* been created, and the name "anObject" now goes to 11. Let's check the id numbers of our two names now, and warm up our `print()` muscles while doing so!

In [15]:
print(f"The ids are {id(anObject)} for anObject \
and {id(notAnotherObject)} for notAnotherObject")

The ids are 140590022007408 for anObject and 140590022008400 for notAnotherObject


---

#### Aside

`print()` is awesome - it is useful for quick & dirty debugging

"f-strings" are also awesome, and relatively new to Python. You make an f-string just like a string, except you put an "f" in front of the opening quote mark.

An f-string lets you directly enclose an expression (like `id(anObject`) in the string enclosed in curly braces, and python will compute the expression, convert it to a string, and pop it right into the f-string!

The backslash ("\"), called the "line continuation character" in this context, tells Python that you want to continue the current line on the next line. You generally do *not* need to use it inside `lists` and such, but you do need it above.

Try running the `print()` above without the `\`.

Now lets make a list (see below), but with each list element on a new line:

In [16]:
aList = [1,
        3, 
        5]
aList

[1, 3, 5]

No "\" needed!

---

Let's make a new variable "eleven" holding the value 11:

In [17]:
eleven = 11

In [18]:
whos

Variable           Type    Data/Info
------------------------------------
aList              list    n=3
anObject           int     11
eleven             int     11
notAnotherObject   int     42


We have now created a few variables in this Python session, so...
**Quick quiz!**

How many names do we currently have?
4 (anObject, notAnotherObject, aList, and eleven)

How many unique objects do we currently have?
3 (anObject and eleven are both 140590022007408)

## Data Types

As we've seen objects in Python have a type (that describes themselves, not the kind of other objects they are attracted to). Later on, we'll learn that we can create our very own custom types but, for now, let's take a look at the common built-in data types.

Remember that all objects you create by naming them will have both *attributes* or *properties* (their nouns) and *methods* (their verbs), but the particular nouns and verbs they have will depend on their type.

Obviously, one thing that varies with type is literally type.

There is a function corresponding to each data type that converts its argument to that data type. The function name is the same as the type as returned by `type()`. Thus, from the "int" from above tells us that there is a function `int()` that will convert things to integers if it can.

### Booleans (bool)

A Boolean variable (named for the mathematician and logician George Boole) can only have the values True and False. 

In [24]:
aBool = True

In [25]:
anotherBool = False

Check the type of these variables.

In [26]:
aBool

True

In [27]:
anotherBool

False

Note that this also tells us that there is a `bool()` function that will convert things to Boolean.

In the cell below, create several Boolean variable, some True, some False.

In [28]:
BoolBoolBool = True

In [30]:
BoolBoolBOOOOL = False

In [31]:
Booleeebooollll = False

In [43]:
BoolBoolBooll = True

Now, in the next code cell, examine the id numbers of the objects corresponding to your names.

In [44]:
print("BoolBoolBool is", BoolBoolBool, "with and id of", id(BoolBoolBool))

BoolBoolBool is True with and id of 4405798176


In [45]:
print("BoolBoolBOOOOL is", BoolBoolBOOOOL, "with and id of", id(BoolBoolBOOOOL))

BoolBoolBOOOOL is False with and id of 4405798904


In [46]:
print("Booleeebooollll is", Booleeebooollll, "with and id of", id(Booleeebooollll))

Booleeebooollll is False with and id of 4405798904


In [47]:
print("BoolBoolBooll is", BoolBoolBooll , "with and id of", id(BoolBoolBooll))

BoolBoolBooll is True with and id of 4405798176


 *In this markdown cell , describe what's going on here in your own words. What's going on with your Boolean variables/objects?*

All the true variables have the same id, and all the false variables also have the same id. This is because the names are for the same variables (true is true and false is false)

Booleans values map to the values of other variables and vice versa. Let's explore this with the two conversion function `int()` and `bool()`.

In the cell below, convert to two possible Boolean values to integers, and convert both zero and some positive and negative non-zero values to Boolean.

In [53]:
print(int(aBool), int(BoolBoolBOOOOL), bool(eleven), bool(notAnotherObject))

1 0 True True


*In this cell, describe the mapping of Booleans to integers and vice versa.*

0 = False, and everything else equals true. So the True boolean value converts to 1, the False boolean value converts to 0, the integer 11 converts to true, and the integer 42 converts to true. The integer 0 converts to False!

In [54]:
itsZero = 0

In [55]:
bool(itsZero)

False

We actually use Boolean values and Boolean logic quite a bit in programming. And, of course, all information on computers is ultimately in the form of bits, so all computers are ultimately Boolean processing machines.

### Integers (int)

Integers are all the numbers without a fractional component which, in practice, means all numbers without a decimal point. As we saw above, creating a variable using an integer automatically creates an object of type `int`.

Create another integer variable now in the cell below and check it's type.

In [56]:
greg = 8

In [57]:
type(greg)

int

Now create another variable with the same value, except put a ".00" after it. So if you integer above was 3, make this new variable equal to 3.00.

In [58]:
gregory = 8.00

Now check the type of your new variable:

In [59]:
type(gregory)

float

### Floating Point Numbers (float)

Floating point numbers are the computer implementation of the real numbers (remember the good ol' number line from school?). They are so named because, unlike a fixed point number, the decimal place is allowed be anywhere in the number, or "float". For example, 1.234, 12.34, and 123.4 are allowed floating point number but, in a fixed point system, only 12.34 would be allowed. In some applications, fixed point numbers have advantages, but we don't have to worry about them, it's just good to know where the name comes from!

Floating point values always have at least 1 digit on either side of the decimal point. Going in other direction, the largest or smallest number you can represent with a `float` (or the largest `int` you can have) is determined by your computer's hardware and operating system.

### Character Strings (str)

Character strings (or just "strings") contain "characters", which are basically the things that you could possible create with your keyboard. 

Strings are enclosed in quotes. Run the code cell below, and this notebook will say hello to you.

In [60]:
aStr = "Hello!"
aStr

'Hello!'

Anything enclosed in quotes is a string, even numbers and operators (like "+").

Make a variable containing a number (e.g. `e_num = 2.7182`), and make another using the same number in quotation marks (e.g. `e_str = '2.7182'`), and check their types:

In [72]:
a_number = 5.678
a_string = '5.678'
type(a_number)
type(a_string)

print("a_number equals", a_number, "and its type is", type(a_number), "a_string equals", a_string, "and its type is", type(a_string))

a_number equals 5.678 and its type is <class 'float'> a_string equals 5.678 and its type is <class 'str'>


Python is string-friendly. Multiple strings can be concatenated just by using the plus sign ("+"). 

Add the string to itself (e.g. `e_str + e_str`):

In [73]:
a_string + a_string

'5.6785.678'

Add the number to itself:

In [74]:
a_number + a_number

11.356

Now try adding the string to the number and vice versa:

In [75]:
a_string + a_number

TypeError: can only concatenate str (not "float") to str

In [76]:
a_number + a_string

TypeError: unsupported operand type(s) for +: 'float' and 'str'

(Note that the errors differ.)

A string is actually the first object we're talking about here that is actually a collection of other things. This means we can get the individual things via *indexing* using square brackets:

In [77]:
aStr[0]

'H'

But what are the things? That is, what are the elements of a string in Python? In the code cell below, see what type the `aStr[0]` (the "H" in "Hello!" is.

In [88]:
type(aStr[0])

str

*Use this cell to explain what this tells you about what multi-element strings are.*

"0" is the first element in aStr (python starts counting at 0). H is the first letter in "Hello". If we did aStr(4), it would show the 5th letter in Hello (o)

In [89]:
aStr[4]

'o'

This tells us that each thing in the string can be like its own object. 

### Tuples (tuple)

Tuples are a collection of objects assigned to a single name. Like this:

In [90]:
myTup = (9, 8, 7)
myTup

(9, 8, 7)

You create it with parentheses (that's what tells Python that you are making a tuple).

In Python, we access the individual items of all multi-element objects using square brackets. Like this:

In [91]:
myTup[0]

9

Recall that we can get multiple things using the colon:

In [92]:
myTup[0:2]

(9, 8)

Tuples can contain objects of different types. Python doesn't care:

In [95]:
myTup = (3, '4', 'five', True)
myTup

(3, '4', 'five', True)

One thing you *can't* do is mess with a tuple. Try changing the 3rd value ('five') to 5:

In [96]:
myTup[2] = 5

TypeError: 'tuple' object does not support item assignment

Why would you want to use a tuple if you can't change the values? If you are working with actual data, for example, you *never* want *anyone* changing the values, so perhaps a tuple is in order.

Also, for large "hungry" applications, tuple are faster and more memory-efficient than lists.

### Lists (list)

Lists are much like tuples except that we can change the individual elements after the list has been created (as much as we want). Lists are created with square brackets. Like this:

In [97]:
myList = [True, 'Hey', 7, 4.5]
myList

[True, 'Hey', 7, 4.5]

Try changing one of these values in the cell below:

In [99]:
myList[0] = False
myList

[False, 'Hey', 7, 4.5]

### Dictionaries (dict)

The most complicated (in some sense) type in Python is the dictionary.

Dictionaries are perhaps poorly named, but they are incredibly useful. A real world dictionary is pretty specific; it yields definitions of words and that's it.

Python dictionaries are much more general and flexible. They allow you to store many elements of any type of data and let you access the data by name. If you've programmed in another language, dictionaries are essentially the same as structs (structures).

Here's a dictionary; notice the curly braces.

In [100]:
myDict = {'name': 'Mary',
         'age': 21,
         'school': 'UT_Austin',
         'gpa': 3.89}

myDict

{'name': 'Mary', 'age': 21, 'school': 'UT_Austin', 'gpa': 3.89}

Dicts are a nice way to store a variety of information about a single logical entity, in this case a hypothetical student named "Mary".

Each entry in the dict is a key:value pair. The keys are strings, and the values can be anything.

We retrieve the information using the keys. Like this:

In [101]:
myDict['name']

'Mary'

Here's a dict to store hypothetical data from an experiment along with "metadata" about the experiment:

In [102]:
expDict = {'date': '10July2023', 
          'time': '1620CST',
          'experimenter': 'FGauss',
          'data': [245.3, 232.9, 238.6, 222.2, 250.1]}

Notice that one of the elements in the dict is a list! If we grab it, it's just a list like any other:

In [103]:
just_a_list = expDict['data']
just_a_list

[245.3, 232.9, 238.6, 222.2, 250.1]

The more you hang around Python, the more you realize how useful dicts are. 

## Mutability

Each type of Python object can either be mutable, meaning that the object itself can be modified, or immutable, meaning that, once created, the object cannot be changed in any way.

### Immutable object types

The common immutable object types are

* bool
* int
* float
* str
* tupple

If you create an immutable object, like a `str`, by running the cell below...

In [110]:
immStr = 'You cannot change me!'
immStr

'You cannot change me!'

You can't change it. Try changing the 3rd element (the element at index 2) to an 'o' to spell 'You' as 'Yoo'.

In [111]:
immStr[3] = 'Yoo'

TypeError: 'str' object does not support item assignment

Consider the following (run the cells in turn):

In [112]:
immStrCopy = immStr

`immStrCopy` and `immStr` both now refer to the same object.

In [113]:
immStr = 'You cannot change me either!'

`immStr` now refers to a newly created object; `immStrCopy` still points to the original unchanged string.

Verify this in the code cell below using the values and ID numbers of the two strings.

In [114]:
print(id(immStr), id(immStrCopy))

140590098578688 140590098577008


### Mutable objects

Mutable objects *can* be changed. This can produce surprising results if you don't appreciate the distinction between mutable and immutable objects.

Common mutable objects are lists and dictionaries.

Consider the following in contrast to the immutable example using strings above.

In [115]:
mList = ['you', 'can', 'change', 'me']
mListCopy = mList

Now change the first (zero index) value of `mList` to 'anyone'. 

In [116]:
mList[0] = 'anyone'

Use the code cell below to check the values and ID numbers of `mList` and `mListCopy`

In [117]:
print(id(mList), id(mListCopy))

140590098581376 140590098581376


In [119]:
mListCopy

['anyone', 'can', 'change', 'me']

*Use this cell to explain how this behavior differs from that of immutable objects.*

For immutable objects, you can't change the objects, so if you copy an object and then change the original, the original and copy are separate things. In a mutable object, you can change the object. So if you copy a mutable object and change the original, both items change and thus have the same ID

## Inquiring minds want to know

There are some functions that allow us to get information about our objects.

### dir()

The `dir()` (for directory) function give us all the verbs for an object.

In the code cell below, make a sting variable containing your name in all lowercase. Then 'dir' it (e.g. `dir(yourNameVar)`).

In [122]:
strName = 'noah'
strName

'noah'

In [126]:
dir(strName)

['__add__',
 '__class__',
 '__contains__',
 '__delattr__',
 '__dir__',
 '__doc__',
 '__eq__',
 '__format__',
 '__ge__',
 '__getattribute__',
 '__getitem__',
 '__getnewargs__',
 '__gt__',
 '__hash__',
 '__init__',
 '__init_subclass__',
 '__iter__',
 '__le__',
 '__len__',
 '__lt__',
 '__mod__',
 '__mul__',
 '__ne__',
 '__new__',
 '__reduce__',
 '__reduce_ex__',
 '__repr__',
 '__rmod__',
 '__rmul__',
 '__setattr__',
 '__sizeof__',
 '__str__',
 '__subclasshook__',
 'capitalize',
 'casefold',
 'center',
 'count',
 'encode',
 'endswith',
 'expandtabs',
 'find',
 'format',
 'format_map',
 'index',
 'isalnum',
 'isalpha',
 'isascii',
 'isdecimal',
 'isdigit',
 'isidentifier',
 'islower',
 'isnumeric',
 'isprintable',
 'isspace',
 'istitle',
 'isupper',
 'join',
 'ljust',
 'lower',
 'lstrip',
 'maketrans',
 'partition',
 'removeprefix',
 'removesuffix',
 'replace',
 'rfind',
 'rindex',
 'rjust',
 'rpartition',
 'rsplit',
 'rstrip',
 'split',
 'splitlines',
 'startswith',
 'strip',
 'swapcase',


Scroll down past all the stuff with double underscores, like `__add__`. The first thing after these should be `capitalize`. Try using this verb using the the dot notation (`object.verb()`).

In [127]:
strName.capitalize()

'Noah'

### %who

Type `who` in the code cell below and run it. (if it doesn't work, try `%who`) 

In [128]:
who

BoolBoolBOOOOL	 BoolBoolBool	 BoolBoolBooll	 Booleeebooollll	 aBool	 aList	 aStr	 a_number	 a_string	 
anObject	 anotherBool	 eleven	 expDict	 greg	 gregory	 immStr	 immStrCopy	 itsZero	 
just_a_list	 mList	 mListCopy	 myDict	 myList	 myTup	 notAnotherObject	 strName	 


*Use this cell to say what `who` does*

'who' tells you all the names you have made, including ones that are assigned to the same object

### %whos

Now try `whos`.

In [129]:
whos

Variable           Type     Data/Info
-------------------------------------
BoolBoolBOOOOL     bool     False
BoolBoolBool       bool     True
BoolBoolBooll      bool     True
Booleeebooollll    bool     False
aBool              bool     True
aList              list     n=3
aStr               str      Hello!
a_number           float    5.678
a_string           str      5.678
anObject           int      11
anotherBool        bool     False
eleven             int      11
expDict            dict     n=4
greg               int      8
gregory            float    8.0
immStr             str      You cannot change me either!
immStrCopy         str      You cannot change me!
itsZero            int      0
just_a_list        list     n=5
mList              list     n=4
mListCopy          list     n=4
myDict             dict     n=4
myList             list     n=4
myTup              tuple    n=4
notAnotherObject   int      42
strName            str      noah


*What does `whos` do?*

'whos' gives you more info about all the names, like their types and the data/info they contain. It also puts them in a nice neat list instead of just listing them messily

Note: who and whos are "magic commands" commands available in iPython (and therefore in Jupyter Notebooks) but are not part of Python *per se*. 